src/theory/datatypes/type_enumerator.cpp

```diff
@@ -166,59 +192,125 @@ Node DatatypesEnumerator::getTermEnum( TypeNode tn, unsigned i ){
    return ret;
  }
 
- void DatatypesEnumerator::init(){
-   //Assert(type.isDatatype());
-   Debug("dt-enum") << "datatype is datatype? " << d_type.isDatatype() << std::endl;
+ void DatatypesEnumerator::init()
+ {
+   Debug("dt-enum") << "datatype is datatype? " << d_type.isDatatype()
+                    << std::endl;
    Debug("dt-enum") << "datatype is kind " << d_type.getKind() << std::endl;
    Debug("dt-enum") << "datatype is " << d_type << std::endl;
-   Debug("dt-enum") << "properties : " << d_datatype.isCodatatype() << " " << d_datatype.isRecursiveSingleton( d_type.toType() );
-   Debug("dt-enum") << " " << d_datatype.isInterpretedFinite( d_type.toType() ) << std::endl;
+   Debug("dt-enum") << "properties : " << d_datatype.isCodatatype() << " "
+                    << d_datatype.isRecursiveSingleton(d_type.toType());
+   Debug("dt-enum") << " " << d_datatype.isInterpretedFinite(d_type.toType())
+                    << std::endl;
 
-   if( d_datatype.isCodatatype() && hasCyclesDt( d_datatype ) ){
-     //start with uninterpreted constant
-     d_zeroCtor = 0;
+   if (d_datatype.isCodatatype() && hasCyclesDt(d_datatype))
+   {
+     // start with uninterpreted constant
      d_has_debruijn = 1;
-     d_sel_types.push_back( std::vector< TypeNode >() );
-     d_sel_index.push_back( std::vector< unsigned >() );
-     d_sel_sum.push_back( -1 );
-   }else{
-     // find the "zero" constructor via mkGroundTerm
+     d_sel_types.push_back(std::vector<TypeNode>());
+     d_sel_index.push_back(std::vector<unsigned>());
+     d_sel_sum.push_back(-1);
+   }
+   else
+   {
+     // find the "zero" term via mkGroundTerm
      Debug("dt-enum-debug") << "make ground term..." << std::endl;
-     Node t = d_type.mkGroundTerm();
-     Debug("dt-enum-debug") << "done : " << t << std::endl;
-     Assert(t.getKind() == kind::APPLY_CONSTRUCTOR);
-     // start with the constructor for which a ground term is constructed
-     d_zeroCtor = datatypes::utils::indexOf(t.getOperator());
+     // Start with the ground term constructed via mkGroundTerm, which does
+     // a traversal over the structure of the datatype to find a finite term.
+     d_zeroTerm = Node::fromExpr(d_datatype.mkGroundValue(d_type.toType()));
```

Comment:
From what I could gather `mkGroundValue` itself seems to depend on the type enumerator... I assume that when you are not well founded this will be captured downstream and this will all break, so this call could throw an exception? I think it'd be good to say it here.

Reply:
I've added more clarification, the argument for non-termination here is independent of this change.